### PR TITLE
Resolve #1718 by no longer calling exec for the find command

### LINKED ISSUE #1718: find command (used to find package:// paths) does not work on windows
in xmlUtil.cpp we do
```
    istringstream p(exec("find -L "+token+" -iname package.xml"));
```
which does not work on windows.
```
dir /b/s 
```
could work on the standard shell, but does not work in the git bash shell, etc.  we need a more robust solution.
boost::filesystem could do it, but we've been avoiding boost.
stl has a filesystem implementation (see http://en.cppreference.com/w/cpp/experimental/fs/recursive_directory_iterator and http://www.technical-recipes.com/2014/using-boostfilesystem/#RecursiveSearch), but it appears to not be available yet on mac clang at least.

### PR DESCRIPTION
**Pull request by sammy-tri**
Incorporates the "tinydir" package to do windows/posix platform traversal.

Not yet tested in windows, so I won't be surprised to see a failure in CI.

### CODE REVIEW

File: drake/thirdParty/tinydir/.travis.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+sudo: false
```
Comment (RussTedrake) [inline drake/thirdParty/tinydir/.travis.yml]: i'm torn as to whether we should have all of the project files from tinydir come into drake.  the copyright file permits modification.  in the extreme, we could keep only the .h file, the COPYING file; it's probably also useful to have the README.md file.  The rest seem like baggage.  But perhaps there are merits to bringing over the entire repo?  What do people think?

File: drake/systems/plants/xmlUtil.cpp [hunk 2]
```diff
@@ -3,19 +3,10 @@
 #include <fstream>
 #include <sstream>
 #include "xmlUtil.h"
+#include "drake/thirdParty/tinydir/tinydir.h"
```
Comment (RussTedrake) [inline drake/systems/plants/xmlUtil.cpp]: btw - i love that you found this solution.  nice, simple, and clean!

**Comment by sherm1** [comment]
> I'm torn as to whether we should have all of the project files from tinydir come into drake. 

I'm more worried about corrupting Drake's exposed external symbols & headers. Since Drake is structured as a library, it will presumably be incorporated into main programs which may themselves be substantial, link in other libraries, and define many symbols. That shouldn't be a problem because Drake should only define symbols in `namespace Drake`. But, if Drake is built with tinydir, TinyXML, etc. and the calling program was also linked to (some version) of those libraries, there will be a conflict. The Debian code police actually check for this (I got in trouble with TinyXML in Simbody). This is a real issue and becoming worse the more great open source code there is to use. For example, see [this issue](https://github.com/USCiLab/cereal/issues/121) in the otherwise-awesome C++ serialization library cereal, which uses an unwrapped version of RapidXML.

I would propose we make it a goal to have Drake guarantee squeaky-clean external symbols and compiler-visible names: `Drake::` for C++ symbols, `Drake_` prefix for C symbols, `DRAKE_` prefix for (hopefully rare) macros.

**Comment by RussTedrake** [comment]
wow.  that hadn't occurred to me.  This one is a header-only library, so should be trivial to encapsulate all of it's symbols in a drake namespace, right?

**Comment by sammy-tri** [comment]
tinydir does take care to define everything as static inline (see the definition for _TINYDIR_FUNC in tinydir.h), so it should bleed between compilation units.  A quick check (on linux, anyway) shows that the symbols are properly local:

**Comment by sherm1** [comment]
> tinydir does take care to define everything as static inline

Oh, that's good -- I didn't notice that. In that case as long as we only include `tinydir.h` in cpp files, and never in Drake headers, those symbols would be hidden.

> On windows the default symbol visibility is hidden anyway

In .dlls, yes, but for .so and .dylib externs would be visible.

One other thing to worry about is namespace-scope class/struct definitions in cpp files. Despite appearances, classes defined in cpp files do not have internal linkage. Two classes in different cpp files that coincidentally have the same name but are not identical can cause non-diagnosed linkage problems (I've been burned by that). I don't know whether structs defined within `extern "C"` have the same problem. Tinydir does define some of those, like `_tinydir_file` and `_tinydir_dir`. Could those cause a problem in a program that was using another version of tinydir in which the structs had changed size? One way to be safe would be to put the struct definitions in an anonymous namespace.

**Comment by sammy-tri** [comment]
Structs defined in extern "C" blocks should not ever emit symbols since (IIRC, it's been a while since I worked in C exclusively) they can't have methods or static member data.  I think the only way we could have a problem with other tinydir-using code is if we tried to pass tinydir_file and tinydir_dir between the two differently compiled bits...  so it should be save to keep using within a single cpp/object file as this PR does.

**Comment by patmarion** [comment]
I'm curious whether tinydir provides API to test whether a path is relative or absolute?  Having that feature would allow me to resolve #1636.

**Comment by sammy-tri** [comment]
It doesn't have any obvious API for such.  The pathname appears to be always composed from the user's input.

**Comment by sherm1** [comment]
@patmarion -- not sure if this is helpful, but Simbody has a more-or-less standalone [Pathname](file:///C:/Users/Sherm/Documents/Binaries/VS14-64/simbody-simbody/html/classSimTK_1_1Pathname.html#details) class that can report relative or absolute correctly on any platform as a side effect of its deconstructPathname() method. The code could easily be extracted.

You might have a different definition of "relative path" in mind. Simbody considers that to be synonymous with "you could reasonably apply a search path to find this file". So "xyz" is relative but "./xyz" is absolute (substitute current directory for "."). On Windows "xyz.txt" is relative but "C:xyz.txt" is absolute (interpreted relative to C drive's current directory).

**Comment by patmarion** [comment]
I actually just need a utility to convert a path (given as a command line argument) to absolute.  Usually that is done like this:

    if not isAboslutePath(path)
        path = join(getCurrentWorkingDir(), path)

I typically use a filesystem library like the one provided by Qt, Python, boost, etc., but drake does not currently have such a library as an available dependency.  Drake use the third party library libspruce, which can get the current working dir and can join paths, but it doesn't have a utility to check whether a path is absolute or relative, which requires a platform dependent implementation.

**Comment by sherm1** [comment]
Is that assuming a smart join? On Windows prepending the current dir isn't sufficient because of the possible presence of the drive letter at the beginning.

**Comment by patmarion** [comment]
Yes, it should be a platform specific join.  I haven't actually checked whether libspruce implements join correctly, I just saw that it provides the API to do so.  I would be happiest to have access to a full featured filesystem manipulation library (which could convert paths for me, so I wouldn't even need to write the implementation I previously posted).  In the past, drake used boost filesystem, but that was phased out some time ago.

**Comment by sherm1** [comment]
Yeah, [this C++17 filesystem](http://en.cppreference.com/w/cpp/experimental/fs) is what we need. Might be good to fake a similar API for the pieces we need to make it easy to throw them away later.

**Comment by rpoyner-tri** [comment]
AFAICT, c++17 filesystem is exactly boost::filesystem, which we can't easily use because of matlab issues. This was discussed already somewhere, but I can no longer find it; my github issue search foo is weak.

**Comment by rpoyner-tri** [comment]
right, billhoffman pointed to the boost BCP tool over in issue #1668.

**Comment by RussTedrake** [comment]
what do people think about the extra files in the PR?  (include travis files, etc).  i'd vote remove them.

**Comment by patmarion** [comment]
agree, just add tinydir.h, that's sorta the point of a light single file library?

**Comment by RussTedrake** [comment]
@sammy-tri -- will wait for the minimal (~ 2 file) version, then accept.

**Comment by sammy-tri** [comment]
retest this please

File: drake/thirdParty/tinydir/.gitignore [hunk 3]
```diff
@@ -0,0 +1,40 @@
+# Object files
```
Comment (RussTedrake) [inline drake/thirdParty/tinydir/.gitignore]: @sammy-tri -- please remove this file.

File: drake/thirdParty/tinydir/.travis.yml (hunk 1 above)
Comment (RussTedrake) [inline drake/thirdParty/tinydir/.travis.yml]: please remove this one, too.

**Pull request merged**